keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,657 @@
+were found.
+    
+A flavor assignment of type 1 means that we need to preempt or wait for other
+workloads in the cohort to finish to accommodate this workload, because the
+ClusterQueue is lending its resources.
+
+A flavor assigment of type 2 means that we need to preempt or wait for other
+workloads in the ClusterQueue to finish to accomodate this workload.
+
+Note that assigments of type 1 and 2 wouldn't borrow quota from the cohort.
+
+Note that a Workload could get flavor assignments of all types.
+
+[#312](https://github.com/kubernetes-sigs/kueue/issues/312) discusses different
+strategies to select a flavor.
+
+#### Sorting Workloads that are heads of ClusterQueues
+
+Sorting uses the following criteria:
+
+1. Flavors that don't borrow first.
+2. [New criteria] Highest priority first.
+3. Older creation timestamp first.
+
+Note that these criteria might put Workloads that require preemption ahead,
+because preemption doesn't require borrowing more resources. This is desired,
+because preemption to recover quota or admit high priority Workloads takes
+preference over borrowing.
+
+#### Admission
+
+When iterating over workloads to be admitted, in the order given by the previous
+section, we disallow borrowing in the cohort in the current cycle after
+evaluating a Workload that doesn't require borrowing. This is the same behavior
+that we have today, but note that this criteria includes Workloads that need
+preemption, because there is no preemption with borrowing quota.
+
+This guarantees that, in future cycles, we can admit Workloads that were not
+heads in their ClusterQueues in this cycle, but could fit without borrowing in
+the next cycle, before lending quota to other ClusterQueues.
+
+In the past, we only disallowed borrowing in the cohort if we were able to
+admit the Workload, because we only kept track of flavor assignments of type 0.
+This caused ClusterQueues in the cohort to continue
+borrowing quota, even if there were pending Workloads that would fit under the
+min Quota for their ClusterQueues.
+
+It is actually possible to limit borrowing only for the involved flavors,
+instead of restricting borrowing for all the flavors in the cohort. But we will
+leave this as a future possible optimization to improve throughput.
+
+#### Preemption
+
+For each Workload that got flavor assignments of type 1 or 2, we might need to
+preempt some admitted Workloads.
+
+The algorithm goes like follows:
+
+0. Check that `ClusterQueue.spec.waitBeforePreemptionSeconds` is satisfied.
+   The time is based on the `lastTransitionTime` of the condition `Admitted`
+   (with `status=False`). If it doesn't exist, we use the current time.
+   If the period is not satisfied, we abort preemption.
+   
+   If this is a ClusterQueue with BestEffortFIFO, the Workload is put in the
+   inadmissible staging area with a deadline equal to the remaining time to
+   fullfil `waitBeforePreemptionSeconds`.
```

Comment:
If no existing workloads can be preempted to fit the incoming workload, then at the limit all workloads that belong to this CQ will always be in the active queue (pretty much converting it into a strict fifo).

Reply:
This is if the deadline is not satisfied. Once it's satisfied, we proceed with preemption and requeue as usual as already explained at the end of this section. Still I tried to rephrase a little.